### PR TITLE
feat(retention) refactor task manager

### LINKED ISSUE #13159: Refactor retention to use the task manager mechanism

### PR DESCRIPTION
**Pull request by bitsf**
refactor retention with task manager
close #13159

### CODE REVIEW

File: src/pkg/retention/controller.go [hunk 1]
```diff
@@ -179,26 +182,52 @@ func (r *DefaultAPIController) DeleteRetention(id int64) error {
...
+	ctx := orm.Context()
+	id, err := r.execMgr.Create(ctx, job.Retention, policyID, trigger,
+		map[string]interface{}{
+			"dry_run": dryRun,
+		},
+	)
 	if _, err = r.launcher.Launch(p, id, dryRun); err != nil {
 		// clean execution if launch failed
-		_ = r.manager.DeleteExecution(id)
+		_ = r.execMgr.Delete(ctx, id)
```
Comment (ywk253100) [inline src/pkg/retention/controller.go:230]: Call  the function `StopAndWait` of execution manager before deleting the execution
The execution cannot be deleted if containing running tasks

File: src/core/service/notifications/jobs/handler.go [hunk 2]
```diff
@@ -91,57 +86,8 @@ func (h *Handler) Prepare() {
...
-	}
-
-	// handle status updating
-	if err := mgr.UpdateTaskStatus(taskID, status, h.revision); err != nil {
-		log.Errorf("failed to update the status of retention task %d: %v", taskID, err)
-		h.SendInternalServerError(err)
-		return
-	}
+	// legacy handler for the retention job
+	return
```
Comment (ywk253100) [inline src/core/service/notifications/jobs/handler.go:90]: We have introduced a common handler to handle the status hooks for the legacy jobs, please refer to https://github.com/goharbor/harbor/blob/master/src/server/route.go#L58

File: src/core/api/base.go [hunk 3]
```diff
@@ -189,9 +190,13 @@ func Init() error {
 
 	retentionMgr = retention.NewManager()
 
-	retentionLauncher = retention.NewLauncher(projectMgr, repository.Mgr, retentionMgr)
+	execMgr := task.NewExecutionManager()
```
Comment (ywk253100) [inline src/core/api/base.go:193]: Use the global execution manager variable instead: https://github.com/goharbor/harbor/blob/master/src/pkg/task/execution.go#L33

The same to the task manager variable

**Comment by ywk253100** [comment]
@bitsf Will you submit another PR to cover the upgrade part?

File: src/pkg/retention/controller.go [hunk 4]
```diff
@@ -179,26 +184,55 @@ func (r *DefaultAPIController) DeleteRetention(id int64) error {
 		}
 	}
 
+	ctx := orm.Context()
```
Comment (wy65701436) [inline src/pkg/retention/controller.go:187]: update the method to give the context as the first parameter?

File: src/pkg/retention/controller.go [hunk 5]
```diff
@@ -179,26 +184,55 @@ func (r *DefaultAPIController) DeleteRetention(id int64) error {
...
+			"dry_run": dryRun,
+		},
+	)
 	if _, err = r.launcher.Launch(p, id, dryRun); err != nil {
 		// clean execution if launch failed
-		_ = r.manager.DeleteExecution(id)
+		if err := r.execMgr.StopAndWait(ctx, id, 10*time.Second); err != nil {
+			logger.Errorf("failed to stop the execution %d: %v", id, err)
+		}
+		_ = r.execMgr.Delete(ctx, id)
```
Comment (wy65701436) [inline src/pkg/retention/controller.go:235]: what about the delete error?

File: src/pkg/retention/launcher.go [hunk 6]
```diff
@@ -75,11 +71,13 @@ type Launcher interface {
 
 // NewLauncher returns an instance of Launcher
 func NewLauncher(projectMgr project.Manager, repositoryMgr repository.Manager,
-	retentionMgr Manager) Launcher {
+	retentionMgr Manager, execMgr task.ExecutionManager, taskMgr task.Manager) Launcher {
```
Comment (wy65701436) [inline src/pkg/retention/launcher.go:74]: do we still need the launcher here? Seems that the methods can be merged into the controller.

File: src/pkg/retention/manager.go [hunk 7]
```diff
@@ -16,17 +16,13 @@ package retention
...
 	"github.com/astaxie/beego/orm"
-	cjob "github.com/goharbor/harbor/src/common/job"
-	"github.com/goharbor/harbor/src/jobservice/job"
 	"github.com/goharbor/harbor/src/pkg/retention/dao"
 	"github.com/goharbor/harbor/src/pkg/retention/dao/models"
 	"github.com/goharbor/harbor/src/pkg/retention/policy"
-	"github.com/goharbor/harbor/src/pkg/retention/q"
 )
 
 // Manager defines operations of managing policy
```
Comment (wy65701436) [inline src/pkg/retention/manager.go:28]: the manager is for policy only here, can we move it to the ./retention/policy?

File: src/pkg/retention/controller.go (hunk 4 above)
Comment (bitsf) [reply src/pkg/retention/controller.go:187]: I will change this in future PR when refactor the API with swagger

File: src/pkg/retention/launcher.go (hunk 6 above)
Comment (bitsf) [reply src/pkg/retention/launcher.go:74]: this retention's launcher still have some special logic not included in task manager

File: src/pkg/retention/manager.go (hunk 7 above)
Comment (bitsf) [reply src/pkg/retention/manager.go:28]: it will introduce some "cycle import"

**Review by ywk253100 — APPROVED** [review]

**Review by wy65701436 — APPROVED** [review]
lgtm

**Pull request merged**